src: introduce process.release object

rvagg opened the pull request:
Replaces #493 and looking for new feedback.

**Requirements**: provide URLs to node-gyp for downloading headers and other required files for building native addons. Decoupling it from a hardwired `http://nodejs.org/dist/` gives us the ability to provide locations for nightlies, next-nightlies, rcs and any other kind of test builds we want to provide with a working node-gyp. This is also the path to stop patching node-gyp for io.js and into the future gives us the option to continue releasing as `iojs` (or whatever name we happened to want!) if that's what we decide.

This new PR simplifies from the original approach, requiring only `--release-urlbase` be passed to `configure`. It's optional but required for release builds (e.g. `make binary RELEASE_URLBASE=https://iojs.org/download/nightly/`). We would supply a string such as `https://iojs.org/download/nightly/`. This gets converted to `process.release.sourceUrl`, `process.release.headersUrl` and `process.release.libUrl` if windows.

A few minor whitespace fixes in Makefile too.

```
$ ./iojs -pe process.release
{ sourceUrl: 'https://iojs.org/download/nightly/v2.3.5-nightly2015071047e2c5c828/iojs-v2.3.5-nightly2015071047e2c5c828.tar.gz',
  headersUrl: 'https://iojs.org/download/nightly/v2.3.5-nightly2015071047e2c5c828/iojs-v2.3.5-nightly2015071047e2c5c828-headers.tar.gz' }
```

_(on Windows you get a `libUrl` property that would point to something like `'https://iojs.org/download/nightly/v2.3.5-nightly2015071047e2c5c828/win-x64/iojs.lib'`)_

thefourtheye commented on configure:
```diff
@@ -675,6 +680,11 @@ def configure_node(o):
     o['variables']['node_tag'] = ''
 
+  if options.release_urlbase:
+    o['variables']['node_release_urlbase'] = options.release_urlbase
```
We can write this simply as `o['variables']['node_release_urlbase'] = options.release_urlbase or ''`. Should we even need the default value? Other options are not doing that.

rvagg replied:
you're right @thefourtheye, it shouldn't be there if it's not set, this is a leftover from a previous incarnation that needed it

rvagg replied:
actually it is required for node.gyp

rvagg commented:
Suggesting that we land this on `next` and do node-gyp work there.

thefourtheye commented:
> gives us the option to continue releasing as `iojs` (or whatever name we happened to want!)

But the windows part in this patch has hard coded `iojs.lib` ;-) Is that okay, if the name we choose is something other than iojs?

rvagg commented:
there's a few things hardcoded as `iojs` just like we hardcode that name in many places, the point is that it's hardcoded here and not in node-gyp which will download whatever is at `libUrl`

Fishrock123 commented:
> Suggesting that we land this on next and do node-gyp work there.

I'm not really sure I see why that would be preferred?

Fishrock123 commented:
SGTM, can't really comment on the c++ part

stevenvachon commented:
Parsing this seems more cumbersome than #493

rvagg commented:
> I'm not really sure I see why that would be preferred?

@Fishrock123 I guess I don't have a good justification, felt right when I said that but can't seem to come up with good reasoning now

@stevenvachon can you elaborate on what you mean by that comment? The main difference between the two is in what needs to be passed to `configure` (and `Makefile` and `vcbuild.bat` for that matter). There were 3 options that needed to go through via config.gypi into the compilation and now I've reduced it to one, making the assumption that it's safe to hard-wire the specific filenames. The actual `process.release` object is the same in both this and #493 except for the missing `process.release.name` property. Or is that the bit that you're concerned about?

thefourtheye commented:
@rvagg Wouldn't it be better if we had the corresponding doc changes as well?

rvagg commented:
@thefourtheye you mean in doc/api? yeah, I'll do them up as well, tests are a bit harder. I'd really like initial feedback from @nodejs/collaborators first to see if there's any significant push-back on this.

Fishrock123 commented:
And, for posterity, `process.release.name` is simply no longer useful in the context of io.js / node.js versioning. There will never be a node.js 1.x or 2.x, as these versions belong to the io.js release line.

stevenvachon commented:
@rvagg Oh, I'd misunderstood both issues. Ultimately, I only need `process.version` now that io.js and node.js have merged.

rvagg commented:
@jasnell @cjihrig additional potential uses for `process.release` for LTS releases:

```
$ ./node -pe process.release
{
  codename: 'Hydrogen',
  supportEnd: {
    LTS: 'Sat Apr 01 2017 00:00:00 GMT+0',
    maintenance: 'Sat Apr 01 2018 00:00:00 GMT+0'
  },
  sourceUrl: '...',
  headersUrl: '...'
}
```

rvagg commented:
I'm going to go ahead and merge this to `master` and also to `next` so I can start testing on a 3.0 RC, which is easier to iterate on that a 2.x release atm. Any objections?

Qard commented:
:+1: I like the supportEnd idea. It'd be cool to have a userland module that could log warnings when rollover periods are reached.

rvagg commented:
derp, I'd better do some docs for this before merging

Fishrock123 commented:
> Any objections?

Hmmm, questions:

1. Downloading the headers still seems messy.
- Why don't we ship with them?
- Should we provide an expected cache dir?
2. What goes in the codename and other LTS fields when it's not an LTS?

rvagg commented:
@Fishrock123 :+1: 

> * Downloading the headers still seems messy.
>    - Why don't we ship with them?

We do ship them, on all but Windows. It's node-gyp's responsibility to pick them up and use them but that's a non-trivial change. We still need to ship the headers for various reasons, mostly about providing a fall-back even when node-gyp is alert to the installed headers.

* Support versions of Node that ship without headers, e.g. Windows, or bad installs, or someone downloading the `iojs.exe` or `node.exe` files we distribute from both iojs.org and nodejs.org.
* `iojs.lib` and `node.lib` are also required for compiles, although they should be shipped with the installer when we get around to fixing that. But they fall into the same category here and are still needed for the same reasons.
* node-gyp supports a mode where it can compile for a specific version of node, not necessarily the one you're running. tbh I don't know if that's used anywhere but that feature would have to be removed if there was no way to get the headers.
* node-gyp will need to have the functionality to download source/headers even if we get smart for future versions, at least for the next few years while we phase out current and past versions of Node. Last I looked, the headers that ship with the 0.12 tarballs can't be used to compile because of some missing components so they are useless even if node-gyp could detect where the header files are!

Plus, we are already downloading the entire source tarball, in gzip form, which takes ages on slower links (like my home connection!). Switching to a header tarball is such a huge leap forward already.

>    - Should we provide an expected cache dir?

Good question, they currently go in to `path.join('~/.node-gyp', process.version)`. When folks were worried that Node.js and io.js versions might overlap it was suggested that we prepend the name of the runtime to that (`'node'` or `'iojs'`). I'm not sure there's a strong need there but perhaps if we wanted the latitude to ship an io.js with an overlapping version number in the future we'd want that. We could achieve it by putting `release.name` back in and letting node-gyp take care of it. I don't think it makes sense either way to spell out exactly what directory it should be because that's a node-gyp (and possible future replacement build tool) decision.

> * What goes in the codename and other LTS fields when it's not an LTS?

_shrug_, empty? I wasn't proposing it for right now, just as a possibility for consideration later so I don't want us to get hung up on this.

jasnell commented:
Oh I like the idea of dropping LTS metadata into this. The only question is
whether it's wise to hard code specific dates...
On Jul 13, 2015 6:44 PM, "Rod Vagg" <notifications@github.com> wrote:

> @jasnell <https://github.com/jasnell> @cjihrig
> <https://github.com/cjihrig> additional potential uses for process.release
> for LTS releases:
>
> $ ./node -pe process.release
> {
>   codename: 'Hydrogen',
>   supportEnd: {
>     LTS: 'Sat Apr 01 2017 00:00:00 GMT+0',
>     maintenance: 'Sat Apr 01 2018 00:00:00 GMT+0'
>   },
>   sourceUrl: '...',
>   headersUrl: '...'
> }
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/nodejs/io.js/pull/2154#issuecomment-121108663>.
>

rvagg commented:
Updated with docs. Also changed the logic so that if `NODE_VERSION_IS_RELEASE` and `--release-urlbase` hasn't been provided it will default to `https://iojs.org/download/release/`. This is so that anyone taking the source tarball to compile a custom version (for example re-packagers for linux distributions) won't have to know & understand how this all works in order to get a build with this metadata.

rvagg commented:
PTAL

rvagg commented:
Also in light of @Fishrock123 raising the issue of cache directory I've added back in `"name": "iojs"` which can be used for a prefix by node-gyp. Also might give some more flexibility around canary releases and version numbering if we needed it.

thefourtheye commented on src/node.cc:
```diff
@@ -2756,6 +2756,39 @@ void SetupProcessObject(Environment* env,
+# endif
+#endif
+
+#if defined(NODE_RELEASE_URLBASE)
```
Since we assign a default value above, we longer need this, right?

rvagg replied:
yes, because it changes depending on release type, that's 1/2 the point here

thefourtheye replied:
I mean, the checking here is not necessary, as `NODE_RELEASE_URLBASE` will definitely be defined at this point.

TooTallNate replied:
Only if `NODE_VERSION_IS_RELEASE` is set.

thefourtheye replied:
@TooTallNate Ah, I overlooked it. Sorry. Thanks for pointing out :-)

jasnell commented:
I'd say let's go ahead and add the `support` metadata support into this now also. Let's move `codename` into `support` and let's please make sure the dates are using ISO8601 interval format and go ahead and omit the time. 

```
$ ./node -pe process.release
{
  name: 'nodejs',
  support: {
    codename: 'Hydrogen',
    active: '2015-10-01/2017-04-01',
    maintenance: '2017-04-01/2018-04-01'
  }
  sourceUrl: '...',
  headersUrl: '...'
}
```

Fishrock123 commented:
> Also in light of Jeremiah Senkpiel raising the issue of cache directory I've added back in "name": "iojs" which can be used for a prefix by node-gyp. Also might give some more flexibility around canary releases and version numbering if we needed it.

I'm not sure that is the best approach, at this point I'd rather not have the name in here.

It would be a better approach to do it as `/dir/node-vx.x.x`. I suppose we don't need to expose this, but it seems like it could be useful for standardizing where they are downloaded and cached to.

rvagg commented:
@jasnell re docs, I agree and have updated, PTAL. Re `support`, we can bikeshed that one when we get closer to LTS but I think we're in rough agreement here (my original version was meant to show proper `Date` objects being stringified).

@Fishrock123 node-gyp has already standardised that for us, providing a full path would be a problem cross-platform I think.

```
$ ls ~/.node-gyp/
0.10.35				1.6.3
0.12.0				1.6.4
1.0.4				1.7.1
1.1.0				2.0.1
1.2.0				2.3.0
1.3.0				2.3.1
1.4.3				2.3.3
1.5.1				3.0.0-rc1
```

I'm suggesting that (a) we leave the exact path the responsibility of node-gyp, or whatever build tool eventually replaces it but (b) we get the name into the path to provide some flexibility.

Another use-case to consider here is extended or alternative node, both of which are slowly becoming reality. The Chakra fork being a good example of this. If they (or perhaps we) ship binaries supporting Chakra then they're going to need headers for Chakra and the shim that makes it look like V8 so that native addons compiled against it do the right thing. They would however have overlapping versions so having `process.binding.name==="node-chakra"` would be helpful, specifically in this case for prefixing the ~/.node-gyp/ subdirectory name to avoid version clashes. JXCore is another example but their add-on compatibility story is even more different given that they also support Spidermonkey. Node-on-JVM is another hypothetical that is not entirely implausible, although the native addon story might be quite a bit more complex for that (some kind of super-NAN?)!

jasnell commented:
@rvagg ... doc changes LGTM

Fishrock123 commented on doc/api/process.markdown:
```diff
@@ -672,6 +672,31 @@ An example of the possible output looks like:
+* `libUrl`: a complete URL pointing to an _iojs.lib_ file matching the
+  architecture and version of the current release. This file is used for
+  compiling add-ons against io.js. _This property is only present on Windows
+  builds of io.js and will be missing on all other platforms._
```
Does not mention `name`?

Fishrock123 commented:
I'm still hesitant about adding `name` on the the point of "what if this devolves into a user-agent -like situation".

Fishrock123 commented:
Ack I guess it's already a `user-agent` situation if people end up having to parse a release URI. LGTM if documented.

rvagg commented:
I've updated this for the v3.0.0-alpha branch @ 62eeeb0f45 (sha not stable), the `name` is `"io.js"` and I've included it on the docs

jasnell commented:
@rvagg ... the new changes LGTM

Fishrock123 commented:
@rvagg could you also push that to here?

rvagg commented:
Sorry @Fishrock123, I screwed up my local version of this and didn't have a fixed branch to push because it was sitting against v3.0.0-alpha. I've fixed it up now and it's updated here.

Fishrock123 commented on doc/api/process.markdown:
```diff
@@ -672,6 +672,33 @@ An example of the possible output looks like:
+
+An Object containing metadata related to the current release, including URLs
+for the source tarball and headers-only tarball. This property may not exist in
+custom builds from non-release versions of the source tree.
```
Won't it still be `{ name: 'io.js' }` or at minimum `{}`?

rvagg replied:
good pickup, thanks!

Fishrock123 commented:
Again. can't comment so much on the C++ or vcbuild stuff, but the rest LGTM minus the docs comment.

Fishrock123 commented:
Landed in 04fd4fad462d6cb6fa3ddfe3e673903935d9cd5a via 3.0.0
